tests/allbackends/base_data_ioarray_prims/Arr.idr

```diff
@@ -0,0 +1,13 @@
+module Arr
+
+import System
+import Data.IOArray.Prims
+
+main : IO ()
+main = do
+  array <- primIO $ prim__newUninitArray 23 {a = String}
+  primIO $ prim__arraySet array 17 "x"
+  "x" <- primIO $ prim__arrayGet array 17
+     | x => do putStrLn (show x)
+               exitFailure
```

Comment:
Can we also add an example where we attempt to read uninitialised memory?
What do we get? An exception? Garbage?

Reply:
> What do we get? An exception? Garbage?

It is implementation dependent. But the Idris convention would be: program may reformat your disk if you read an element before writing it.

In RefC the array is initialized with `NULL`, so reading initialized memory will result in sevg when accessing contents of that element.

It should be something similar for Node.

For chez I don't really know, the specification says "elements are unspecified", when I tried to read, it returned zero, but it could be anything else.

> Can we also add an example where we attempt to read uninitialised memory?

Example where?

Reply:
In the test case. I was hoping we could catch an exception if it was reliably crashing
or simply ignore the result if it was reliably returning garbage. But if the result is
unpredictable then we can't do much.

Just to double check: if you add `%default total` in the test file, it does complain
about the call to `prim__newUninitArray`, right?

Reply:
> In the test case.

It is not possible to write such test case, because reading may crash.

> I was hoping we could catch an exception if it was reliably crashing or simply ignore the result if it was reliably returning garbage.

These are low-level primitives, building blocks for safer primitives like `IOArray`. `IOArray` guarantees that we don't return trash, and in particular, we don't read uninitialized memory.

> Just to double check: if you add %default total in the test file, it does complain about the call to prim__newUninitArray, right?

I need to check, but I believe it doesn't complain: all `prim__` array functions are declared total.

Reply:
> all `prim__` array functions are declared total.

Right but they are because of initialization and bound checks.
This is different: we're introducing a source of partiality and
it should be tagged as such. If it's locally used in a safe manner
(to implement `fromList` for instance) then it can be wrapped in
an `assert_total` locally.

Reply:
But the same applies to other three functions:

* `newArray` may fail on memory allocation, so technically it is not total according to strict definition of totality
* `arrayGet` and `arraySet` may fail because of index being out of range

Maybe we want to declare all functions partial, I can submit a PR to do that. But for new function I prefer symmetry with existing functions.